openstack/baremetal/v1/nodes/requests.go

```diff
@@ -301,3 +320,118 @@ func Delete(client *gophercloud.ServiceClient, id string) (r DeleteResult) {
 	_, r.Err = client.Delete(deleteURL(client, id), nil)
 	return
 }
+
+// A cleaning step has required keys ‘interface’ and ‘step’, and optional key ‘args’. If specified,
+// the value for ‘args’ is a keyword variable argument dictionary that is passed to the cleaning step
+// method.
+type CleanStep struct {
+	Interface string            `json:"interface,required"`
+	Step      string            `json:"step,required"`
+	Args      map[string]string `json:"args,omitempty"`
+}
+
+// ProvisionStateOptsBuilder allows extensions to add additional parameters to the
+// ChangeProvisionState request.
+type ProvisionStateOptsBuilder interface {
+	ToProvisionStateOpts() (map[string]interface{}, error)
+}
+
+// ProvisionStateOpts for a request to change a node's provision state.
+type ProvisionStateOpts struct {
+	Target         TargetProvisionState `json:"target,required"`
+	ConfigDrive    ConfigDriveBuilder   `json:"-"`
+	CleanSteps     []CleanStep          `json:"clean_steps,omitempty"`
+	RescuePassword string               `json:"rescue_password,omitempty"`
+}
+
+// ConfigDriveBuilder allows extensions to change how the configdrive is constructed.
+type ConfigDriveBuilder interface {
+	ToConfigDrive() (*string, error)
+}
+
+// A configdrive may be a path to a directory, or an image file
+type ConfigDrivePath string
+
+// A configdrive may also be a URL pointing to such an image
+type ConfigDriveValue string
+
+// Converts a path to a valid config drive.  If a directory, will be packed as an ISO and gzipped.  If a file,
+// will be gzipped if not.  The final result is base64 encoded.  The return is a pointer to avoid potentially copying
+// a large value.
+func (path ConfigDrivePath) ToConfigDrive() (*string, error) {
```

Comment:
Config drives are used to support metadata for configuration the bare metal host. They are ISO9660 images, and when sent to the API need to be gzipped and base64 encoded.  I'm not sure best how to handle this in gophercloud, so I went the route of taking a bunch of different options and doing the right thing. I'm not sure where to draw the line though in how much gophercloud should do.  This PR handles several cases:

1. If a user gives us a directory, it'll be turned into an ISO image, gzipped, and base64 encoded
2. If the user gives us an ISO image, it'll be gzipped and base64 encoded
3. If the user gives us a gzipped file, it'll be base64 encoded
4. If the user gives us a URL, we'll send that unmodified

Reply:
This is a tough one.

The easiest decision is about detecting ISO, directory, etc - this level of support shouldn't be in Gophercloud. The reason is because Gophercloud is too low-level to implement such decisions. For example, you're using `mkisofs` to create an ISO which kind of says `mkisofs` is the official solution and creates the possibility of having to field support because some people want to use `mkisofs` differently but don't want to create a new OptsBuilder etc etc 😄 

Once that's removed, that leaves us with: the value can either be a base64 encoded string or a URL.

This seems like an odd rule, no? These are two vastly different types of data being overloaded in the same request parameter. Is Ironic actually checking if the value is one or the other? Or could you use a value of `foobar` and Ironic would happily use it?

Either way, because of the overloading, I'm leaning on having the `ConfigDrive` field simply be a string and pushing all validation off to the application. 

It's possible to detect if the value is a URL by trying to parse it as a URL in the `ToProvisionStateMap` method. But then it gets weird if it's not a URL - how far should we go? Check if it's a valid base64 string? Check if it has gzip encoding? unzip it and see if it's a valid ISO9660 encoding? And what if the string is 500mb in size (more on that later)?

Deferring to the application makes a little more sense. For example, if I were to implement this in Terraform, I could have a few end-user facing arguments:

* `configdrive_url`
* `configdrive_iso`
* `configdrive_dir`

Each would be mutually exclusive to each other, each would have their own Terraform-specific validation rules applied, and each would be handled in a separate way. However, the final result of all 3 would be either a URL string or a base64 gzipped string.

There's one caveat to this: is there a limit to how large the `configdrive` parameter can be? If not, this _really_ complicates things. Imagine someone having a 500mb gzip'd base64 string. That's going to cause memory exhaustion in Go. To handle this, the `ConfigDrive` field is going to need to be an `io.Reader` rather than a string, and the `ChangeProvisionState` function is going to look similar to the object storage [`Create`](https://github.com/gophercloud/gophercloud/blob/master/openstack/objectstorage/v1/objects/requests.go#L222) function which uses [`RawBody`](https://github.com/gophercloud/gophercloud/blob/master/openstack/objectstorage/v1/objects/requests.go#L240) to stream bytes. This would allow an ISO to be read from disk and be streamed to Ironic with little or no buffering, thus keeping memory in check.

No doubt there is more to discuss here. I don't have a good understanding of how the Ironic API works, so I could be wrong about much of the above.

Reply:
> The easiest decision is about detecting ISO, directory, etc - this level of support shouldn't be in Gophercloud. The reason is because Gophercloud is too low-level to implement such decisions. For example, you're using mkisofs to create an ISO which kind of says mkisofs is the official solution and creates the possibility of having to field support because some people want to use mkisofs differently but don't want to create a new OptsBuilder etc etc smile

Nothing is preventing someone from using a different ISO tool, but if they want to pass a directory and have us do it, then it's obviously has to be done in an opinionated way. I know you picture gophercloud being low-level but I'm not sure of it's benefit over making API calls myself if it's not adding value. Anyone using Ironic is probably provisioning bare metal nodes that need node-specific customizations and will need this functionality. If that's still not convincing I'll happily remove it and we'll have to maintain similar code in metalkube (my personal target for using gophercloud), and likely everyone else using Ironic will need to do something similar.

FWIW, mkisofs is portable and basically the way everyone does things like this, for example Terraform uses it and it's available on most platforms (e.g. it's in brew on os x).

> This seems like an odd rule, no? These are two vastly different types of data being overloaded in the same request parameter. Is Ironic actually checking if the value is one or the other? Or could you use a value of `foobar` and Ironic would happily use it?

There's no validation on the Ironic side from what I can see, it gets passed to Swift behind the scenes and it ostensibly should complain about foobar not being a file or URL which would bubble back up to us as an error.

> Check if it's a valid base64 string? Check if it has gzip encoding? unzip it and see if it's a valid ISO9660 encoding?

If the user passes a string value, then we just assume they did the right thing by putting a URL or base64 encoding something themselves.  But if they're giving us a file name, can we check if it's gzipped? Or should we not even do that?

> There's one caveat to this: is there a limit to how large the configdrive parameter can be?

There is, 64mb and in practice much smaller - a few dozen kb.  It's generally just small text files (e.g. kickstarts, ignition for CoreOS, etc).

Reply:
> I know you picture gophercloud being low-level but I'm not sure of it's benefit over making API calls myself if it's not adding value.

I think that kind of reaction is a sign that something has been designed well.

Gophercloud handles authentication, pagination, field types, request and response structs, errors, and things like that - it's doing a lot of work on top of a generic [http request](https://golang.org/pkg/net/http/#Client).

Gophercloud _could_ handle things like running mkisofs, but there needs to be a limit. Given that there's no other area in Gophercloud that requires a subcommand to be run, I really need to question if it's _absolutely_ necessary. This isn't personal by any means.

> I'll happily remove it and we'll have to maintain similar code in metalkube (my personal target for using gophercloud), and likely everyone else using Ironic will need to do something similar

The more I think about it, the more it sounds like `ConfigDrive` should just be a plain string without validation.

I agree with your view on repetitive code which would need to be repeated in every app supporting Ironic via Gophercloud, so I think the best thing to do is to make a utility function at https://github.com/gophercloud/utils which handles this. 

The utility function could be called the same (`ChangeProvisionState`) and simply be a wrapper around the "core" `ChangeProvisionState`. 

But this is where you could do things like hard-code `mkisofs`, read in a series of files, etc. You could even create a more user-friendly struct for handling the different ways a config drive could be constructed.

It sounds nuanced, but it provides a clear distinction of roles as well as direction (app consumes utils which consumes gophercloud).

Thoughts?

Reply:
Commented above, but will repeat here: https://storyboard.openstack.org/#!/story/2005083 solves at least some of these concerns, if it gets accepted.

Reply:
To be clear: if this change is accepted, it will be guarded by a microversion. So some optional solution will be needed for older servers.

Reply:
That does make things a lot easier. 

For now I think it’s reasonable to restore the value to a string and let the user do what’s needed. I’ll move the code to help making the ISO to the util repo, thanks!